### PR TITLE
GH-1153 Adds new PropertyEditor

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes GH-1153

@GordonSmith @mlzummo @dtsnell4 Please review

http://rawgit.com/jbrundage/Visualization/Flat_Prop_Form/demos/dermatology.html

Signed-off-by: Jay Brundage <jaman.brundage@lexisnexis.com>

### CODE REVIEW

**Comment by jbrundage** [comment]
@GordonSmith This PR should resolve the slowness experienced in DSP and I believe it's ready to be pulled, could you review?

**Comment by mlzummo** [comment]
@jbrundage we can test this tomorrow on that slow WU in DSP we'll work on it in integration

**Comment by GordonSmith** [comment]
MultiChart has the issue when you switch the chart type the child properties aren't refreshed (clicking on the sort annotation forces a refresh) - adding the watch back in should fix it and with your merging of data now, lots of updates calls will be cheap when nothing changes!

**Comment by jbrundage** [comment]
Good point, I'll add the watch back in first thing in the morning.

**Comment by jbrundage** [comment]
@GordonSmith I've reintroduced the watch feature. And if you toggle the "Property Tree" checkbox at the top of the page you can view the hierarchy of currWidget's publish params in a CirclePacking widget.

PS: We need a better way to determine the placement of **main** cells on dermatology.html

File: src/other/PropertyEditor.js [hunk 1]
```diff
@@ -1,401 +1,266 @@
...
+        var context = this;
         
-        if(this._parentElement.filter(".other_PropertyEditor div").empty()){
-            this._parentElement.style("overflow", "auto");
-        }
+        if(!this.widget())return;
         
-        if(this.title() === ""){
-            this.title("Property Editor");
+        this.setWatches();
```
Comment (GordonSmith) [inline src/other/PropertyEditor.js]: Should only be called once per widget (not during update)?

File: src/other/PropertyEditor.js [hunk 2]
```diff
@@ -1,401 +1,266 @@
...
-        };
-        var ret = {};
-        for(var p in arr){
-            for(var c in categories){
-                if(p.toLowerCase().indexOf(categories[c]) !== -1){
-                    break;
+    PropertyEditor.prototype.setWatches = function (domNode, element) {
+        var context = this;
+        context.widget().watch(function(paramId,newVal,oldVal){
+            if(context._flatParams[paramId] instanceof Array){
```
Comment (GordonSmith) [inline src/other/PropertyEditor.js]: If any watched field is changed, then a simple call to render() should be made (as this is no longer an expensive exercise)?
Specifically I am looking at multi chart and changing the outer "data" - the inner data for the chart is not being updated (but it does if I click the sort)...

**Comment by GordonSmith** [comment]
http://rawgit.com/jbrundage/Visualization/Flat_Prop_Form/demos/dermatology.html?src/chart/MultiChart:
image

File: src/other/PropertyEditor.js (hunk 2 above)
Comment (jbrundage) [inline src/other/PropertyEditor.js]: I've locally modified the watch callback to simply call context.render(), and it has become a performance issue. I'm working on it now, the watches for child widgets are likely not being removed

Comment (GordonSmith) [inline src/other/PropertyEditor.js]: I am hoping that once your not leaking watches it will be fast enough to re-render when any value changes?

**Comment by jbrundage** [comment]
@GordonSmith The watch leaking has been resolved... and now the watch callback runs context.render()

**Comment by GordonSmith** [comment]
And it passed my sneaky test:
image
(and is fast)
I will review the code once I get a chance...

File: demos/dermatology.html [hunk 3]
```diff
@@ -172,13 +177,16 @@
...
         var testTheme = null;
         var testWidget = null;
         var currWidget = null;
+        var propEditor = null;
         var showProperties = null;
+        var showPropertyTree = null;
         var debouncedResize = null;
 
-        require(["d3", "src/layout/Surface", "src/layout/Grid", "src/other/Persist", "src/other/PropertyEditor", "test/Factory"], function (d3, Surface, Grid, Persist, PropertyEditor, testFactory) {
+        require(["d3", "src/layout/Surface", "src/layout/Grid", "src/other/Persist", "src/other/PropertyEditor", "src/tree/CirclePacking", "test/Factory"], function (d3, Surface, Grid, Persist, PropertyEditor, Dendrogram, testFactory) {
```
Comment (GordonSmith) [inline demos/dermatology.html]: Should this be Dendrogram?

Comment (jbrundage) [inline demos/dermatology.html]: The Dendrogram has major issues with overlapping text, but CirclePacking visualizes the hierarchy nicely.

**Comment by GordonSmith** [comment]
When showing the CiriclePacking view you are being hit by the Grid Multiplier issue:  
image

File: demos/dermatology.html [hunk 4]
```diff
@@ -253,10 +263,16 @@
...
-                    ;
+                    propEditor.widget(sourceWidget);
+                    propEditor.render();
+                }
+            }
+            
+            function displayPropertyTree(sourceWidget) {
+                if (document.getElementById("showPropertyTree").checked) {
+                    sourceWidget = sourceWidget || currWidget;
+                    propTree.data(propEditor.getDendrogramObject());
```
Comment (GordonSmith) [inline demos/dermatology.html]: Rename "getDendrogramObject" "getDataTree" (I am trying to convey that you are returning the data structure as a ITree compatible data structure)?

File: demos/dermatology.html [hunk 5]
```diff
@@ -253,10 +263,16 @@
...
+                    propTree.render();
```
Comment (GordonSmith) [inline demos/dermatology.html]: Minor:
```
propTree
  .data(...)
  .render()
;
```

File: demos/dermatology.html [hunk 6]
```diff
@@ -346,6 +364,13 @@
                     })
                 ;
             };
+            showPropertyTree = function (show) {
+                propTree.data(propEditor.getDendrogramObject());
+                main
+                    .setContent(3, 0, show ? propTree : null, "", 5, 3)
```
Comment (GordonSmith) [inline demos/dermatology.html]: Your row position is wrong due to the internal cell multiplier.

File: src/form/PropertyInput.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                    });
+            }
+        });
+    }
+
+    return {
+        appendInputs: appendInputs,
+        updateInputs: updateInputs,
+    };
+}));
```
Comment (GordonSmith) [inline src/form/PropertyInput.js]: Are you planning on getting re-use out of this file (I am not sure you will be able to?), if not I would add those two functions back into the PropertyEditor source file?

**Comment by jbrundage** [comment]
That white-space is there to allow space for the S. Test to display at the same time. Ultimately dermatology.html needs some logic to auto-magically handle Cell placement.

File: src/form/PropertyInput.js (hunk 7 above)
Comment (jbrundage) [inline src/form/PropertyInput.js]: The idea was to keep the input code separate from the rest of the PropertyEditor code. I'll merge the two files.

**Comment by GordonSmith** [comment]
Closing in favor of:  https://github.com/hpcc-systems/Visualization/pull/1213

**Pull request closed without merging**